### PR TITLE
NIAD-757: Move /DocumentReference parameter validation to external va…

### PR DESCRIPTION
**Pull request by graham-andrew**
…lidator class

## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            checkGetScrTypeParam(type);
+        } catch (BadRequestException exc) {
+            setErrorMessage(context, exc.getMessage());
+            return false;
+        }
+
+        return true;
+    }
+
+    private void checkGetScrTypeParam(String type) {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java:35]: I don't think there's a point in exception-driven approach here. It's just one check - why not return boolean?
This applies to all new validators.

**Review by dawidskainos — CHANGES_REQUESTED** [review]

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java (hunk 1 above)
Comment (graham-andrew) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java]: Changed validation checks to return booleans.

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/nhsnumber/PatientIdParameter.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Constraint(validatedBy = PatientIdParameterValidator.class)
+@Target(PARAMETER)
+@Retention(RUNTIME)
+public @interface PatientIdParameter {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/nhsnumber/PatientIdParameter.java:14]: Let's change class name to PatientId

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/count/RecordCountParameter.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Constraint(validatedBy = RecordCountParameterValidator.class)
+@Target(PARAMETER)
+@Retention(RUNTIME)
+public @interface RecordCountParameter {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/count/RecordCountParameter.java:14]: Let's change class name to RecordCount

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/count/RecordCountParameterValidator.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class RecordCountParameterValidator implements ConstraintValidator<RecordCountParameter, Integer> {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/count/RecordCountParameterValidator.java:12]: Let's change class name to RecordCountValidator

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/nhsnumber/PatientIdParameterValidator.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.stereotype.Component;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class PatientIdParameterValidator implements ConstraintValidator<PatientIdParameter, String> {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/nhsnumber/PatientIdParameterValidator.java:14]: Let's change class name to PatientIdValidator

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/sort/SortMethodParameter.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Constraint(validatedBy = SortMethodParameterValidator.class)
+@Target(PARAMETER)
+@Retention(RUNTIME)
+public @interface SortMethodParameter {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/sort/SortMethodParameter.java:14]: Let's change class name to SortMethod

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/sort/SortMethodParameterValidator.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.stereotype.Component;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class SortMethodParameterValidator implements ConstraintValidator<SortMethodParameter, String> {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/sort/SortMethodParameterValidator.java:14]: Let's change class name to SortMethodValidator

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/sort/SortMethodParameterValidator.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    public boolean isValid(String sort, ConstraintValidatorContext context) {
+        if (!checkGetScrSortParam(sort)) {
+            setErrorMessage(context, String.format("Invalid value - %s in field '_sort'", sort));
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean checkGetScrSortParam(String sort) {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/sort/SortMethodParameterValidator.java:32]: Let's use is prefix

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameter.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Constraint(validatedBy = TypeCodeParameterValidator.class)
+@Target(PARAMETER)
+@Retention(RUNTIME)
+public @interface TypeCodeParameter {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameter.java:14]: Let's change class name to TypeCode

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.stereotype.Component;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class TypeCodeParameterValidator implements ConstraintValidator<TypeCodeParameter, String> {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java:14]: Let's change class name to TypeCodeValidator

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    public boolean isValid(String type, ConstraintValidatorContext context) {
+        if (!checkGetScrTypeParam(type)) {
+            setErrorMessage(context, String.format("Invalid value - %s in field 'type'", type));
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean checkGetScrTypeParam(String type) {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java:32]: let's use is prefix

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/nhsnumber/PatientIdParameterValidator.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    public boolean isValid(String patientId, ConstraintValidatorContext context) {
+        if (!checkNhsNumber(patientId)) {
+            setErrorMessage(context, String.format("Invalid value - %s in field 'patient'", patientId));
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean checkNhsNumber(String patientId) {
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/nhsnumber/PatientIdParameterValidator.java:32]: let's use is prefix

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
+package uk.nhs.adaptors.scr.controllers.validation.getscrid.type;
```
Comment (dawidskainos) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/validation/getscrid/type/TypeCodeParameterValidator.java:1]: let's put all scr-id validation classes in validation/scr package

**Review by dawidskainos — CHANGES_REQUESTED** [review]

**Review by dawidskainos — APPROVED** [review]

**Pull request merged**